tez-tests/src/test/java/org/apache/tez/test/TestAMRecoveryAggregationBroadcast.java

```diff
@@ -0,0 +1,513 @@
+
+    assertEquals(Collections.emptyList(), readRecoveryLog(3));
+  }
+
+  /**
+   * v1 scans lines and emit (line -> 1), imitating a simple Map vertex.
+   * v2 aggregates the lines and emit (line -> # of duplicated values), imitating an aggregation.
+   * v3 joins the output of v2 with another input. v2 broadcasts its output.
+   * (input1)
+   *    \
+   *    v1
+   *     \
+   *     v2  (input2)
+   *      \  /
+   *       v3
+   */
+  private DAG createDAG(String dagName) throws Exception {
+    UserPayload payload =TezUtils.createUserPayloadFromConf(tezConf);
+    DataSourceDescriptor dataSource = MRInput
+        .createConfigBuilder(new Configuration(tezConf), TextInputFormat.class,
+            INPUT_FILE.toString())
+        .build();
+    // each line -> 1
+    Vertex tableScanVertex = Vertex
+        .create(TABLE_SCAN, ProcessorDescriptor.create(TableScanProcessor.class.getName())
+            .setUserPayload(payload))
+        .addDataSource(INPUT1, dataSource);
+
+    // key -> num keys
+    Vertex aggregationVertex = Vertex
+        .create(AGGREGATION, ProcessorDescriptor
+            .create(AggregationProcessor.class.getName()).setUserPayload(payload), 1);
+
+    DataSinkDescriptor dataSink = MROutput
+        .createConfigBuilder(new Configuration(tezConf), TextOutputFormat.class,
+            OUT_PATH.toString())
+        .build();
+    // Broadcast Hash Join
+    Vertex mapJoinVertex = Vertex
+        .create(MAP_JOIN, ProcessorDescriptor.create(MapJoinProcessor.class.getName())
+            .setUserPayload(payload))
+        .addDataSource(INPUT2, dataSource)
+        .addDataSink(OUTPUT, dataSink);
+
+    EdgeProperty orderedEdge = OrderedPartitionedKVEdgeConfig
+        .newBuilder(Text.class.getName(), IntWritable.class.getName(), HashPartitioner.class.getName())
+        .setFromConfiguration(tezConf)
+        .build()
+        .createDefaultEdgeProperty();
+    EdgeProperty broadcastEdge = UnorderedKVEdgeConfig
+        .newBuilder(Text.class.getName(), IntWritable.class.getName())
+        .setFromConfiguration(tezConf)
+        .build()
+        .createDefaultBroadcastEdgeProperty();
+
+    DAG dag = DAG.create("TestAMRecoveryAggregationBroadcast_" + dagName);
+    dag.addVertex(tableScanVertex)
+        .addVertex(aggregationVertex)
+        .addVertex(mapJoinVertex)
+        .addEdge(Edge.create(tableScanVertex, aggregationVertex, orderedEdge))
+        .addEdge(Edge.create(aggregationVertex, mapJoinVertex, broadcastEdge));
+    return dag;
+  }
+
+  TezCounters runDAGAndVerify(DAG dag, boolean killAM) throws Exception {
+    tezSession.waitTillReady();
+    DAGClient dagClient = tezSession.submitDAG(dag);
+
+    if (killAM) {
+      TimeUnit.SECONDS.sleep(10);
+      YarnClient yarnClient = YarnClient.createYarnClient();
+      yarnClient.init(tezConf);
+      yarnClient.start();
+      ApplicationAttemptId id = ApplicationAttemptId.newInstance(
+          tezSession.getAppMasterApplicationId(), 1);
```

Comment:
nit: line break is not needed

Reply:
Done
https://github.com/apache/tez/pull/361/commits/7f0b3b8554ab26d481c10f665ba5ee7881135e68